New Grid Filter UX Text Automation.

labkey-danield opened the pull request:
#### Rationale
This updates the test cases in GridPan

<img width="901" alt="Screen Shot 2022-05-23 at 9 14 27 PM" src="https://user-images.githubusercontent.com/12738200/169947617-f43ab7f6-6f95-45d2-bfbd-10c289ecb309.png">
elTest to test more functionality in the new grid filter dialog.

<img width="901" alt="Screen Shot 2022-05-23 at 9 14 40 PM" src="https://user-images.githubusercontent.com/12738200/169947627-bc42db59-77c8-4058-a65e-880699f8dd53.png">

The [New Grid Filter UX spec](https://docs.google.com/document/d/1PyPw34qpInoicg-hDx0Qzw19PmTrTi6A2s-c2FwMhAI/edit#) identified several test scenarios the needed to be covered. In addition several other tests were added as well to cover other gaps. As a result there are a lot of changes to the GridPanelTest, and it might be easier to view the file as if it were a new file and not as a diff. One of the biggest changes to the test was to use a shared sample type. Previously each test was creating its own sample type, populating and then deleting it at the end of the test. Several of the  older tests were consolidated into other tests.

For the most part only minor changes were made to the test components. Most of the changes involve how to access the filter dialog (e.g. clicking a filter 'pill' above the grid now returns the dialog). Other changes involved fixes to test components.

#### Related Pull Requests
* TBD

#### Changes
* Updated and consolidate tests in GridPanelTest.
* Removed calls to deprecated methods from GridPanelTest.
* Updated Pager component to use MultiMenu (removed references to deprecated DropDownButtonGroup)

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Remove calls to deprecated methods and other linter suggestions.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Use a shared sample type int he tests (don't create a sample type in each test).
Remove redundant tests.
Add new grid panel tests.
Remove calls to deprecated methods.
Add method to check for pagination controls.
Add method to check for 'Select All' button.
Fix getter for button text in MultiMenu
Add getter for filter dialog to grid bar.

labkey-danield committed:
Reverting get button text.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Fixed some finders.
Fixed issue when clicking a tab.
Added a waitForUpdate when using the search box on a grid.
Changed the panel getters in the GridFilterDialog to invalidate the cache when the panel is loaded.
Added getter for tab text.
Fixed getter for current page size from the pager.
Added more tests.

labkey-danield committed:
Updating 'pill' test to have multiple filters on one column.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Added search tests.
Added extended characters test.

labkey-danield committed:
Added export tests.
Add helper to query tables to get expected results (to validate filter results).
Add custom views to sample type during test setup.

labkey-danield committed:
Added tests with custom views.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Add tests for LKSM sources.
Add test for box display.
Update Tabs to not persist grid panel (find them every time).

cnathe commented on src/org/labkey/test/tests/component/GridPanelTest.java:
```diff
@@ -1,385 +1,1749 @@
+     *
+     * @return True if the dropdown page menu is visible, false otherwise.
+     */
+    public boolean isPagingMenuVisible()
```
should this helper be on the gird component itself?

labkey-danield replied:
I wasn't sure if it would be useful outside this test, but I can see that it would make sense to be in a component. I've moved it to the Pager test component.

cnathe reviewed (commented):
Test cases in the refactored GridPanelTest look good! I'll defer the rest of the code review to the test team. Thanks Dan.

labkey-danield committed:
Moving isPagingMenuVisible method out of the tests and into the Pager object.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/GridBar.java:
```diff
@@ -416,15 +416,24 @@ else if(onSamplePage)
     {
-        elementCache().searchBox.set(searchStr);
-        elementCache().searchBox.getComponentElement().sendKeys(Keys.ENTER);
+        _queryGrid.doAndWaitForUpdate(()->
+        {
+            elementCache().searchBox.set(searchStr);
+            elementCache().searchBox.getComponentElement().sendKeys(Keys.ENTER);
+        });
```
I think there was a similar change to `searchFor` in the last update to this component and it had to get rolled back. The grid wont update if the search box doesn't actually change (I think this needs something similar to the `if` in `clearSearch`)

labkey-danield replied:
As discussed, added call to clearSearch before setting the new value.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/GridBar.java:
```diff
@@ -433,6 +442,12 @@ public String getSearchExpression()
         return elementCache().searchBox.get();
     }
 
+    public GridFilterModal getFilterDialog()
+    {
+        clickButton("Filters");
+        return new GridFilterModal(getDriver(), _queryGrid);
+    }
+
```
`get` methods generally shouldn't change the page state. Maybe `openFilterDialog` instead.

labkey-danield replied:
Changed.

labkey-tchad reviewed (approved):
This might be reintroducing a defect in `searchFor`

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Changing function name.
Change code for search textbox to clear it before entering a new value.
Added test to check filter with a custom view that is only a filter.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest

labkey-danield committed:
Added verifyEqualsSorted to DeferredErrorCollector.
Modified new tests to use verifyEqualsSorted.

labkey-danield committed:
Update version of grid control used.

labkey-danield committed:
Merge branch 'develop' into fb_additionalGridPanelTest
